### PR TITLE
Upgrade Android SDK to API level 17

### PR DESCRIPTION
**Pull request by badboy**
This is a followup to the [call for help](https://internals.rust-lang.org/t/help-wanted-segfault-in-android-stdlib-tests-default-plan-to-disable-tests-on-android/12058).

It should upgrade the Android API level Rust is compiled against from 14 to 17.

* #70958 disabled some tests that might have been the culprit of frequent test crashes
* According to [this comment](https://github.com/rust-lang/rust/issues/55861#issuecomment-609638213) API level 17 is Android 4.2, released ~8 years ago.
* Just last week [there were dashboards](https://developer.android.com/about/dashboards/) about Android usage. Android <4 was well below 1%.

For now all I did was replace `14` with `17` in the Dockerfiles, fix some comments, remove special-cased functions and fixed some imports.

Tested with the `arm-android` Docker image. [Full log in a Gist](https://gist.github.com/badboy/be2e5c9fcf126bcbebad05644eca3c9d).
~~I don't actually know if the upgrades work just like that or if it requires someone from infra to help out.~~

~~If this all _does_ work out we could consider re-enabling the tests from #70958 to see how it goes.~~

---

Revisions:

* The first round of this PR increased only to API level 16, but that would not have solved the [`posix_memalign` issue](https://github.com/rust-lang/rust/pull/71123#issuecomment-623978943), so I opted to go one higher.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
cc @joshtriplett for tiering policy

**Comment by badboy** [comment]
Happy to update the PR, once I get information if this approach is indeed the right one.

**Comment by Mark-Simulacrum** [comment]
cc @rust-lang/compiler @rust-lang/release -- do we know of any reason *not* to do this? It looks reasonable to me given the usage figures and such.

@badboy -- could you enable to dist-android and arm-android on the "try" builder and then we can run that here to check if your changes work?

Something like this should work:

```diff
diff --git a/src/ci/azure-pipelines/try.yml b/src/ci/azure-pipelines/try.yml
index 38a0685e0f7..6244e1413d0 100644
--- a/src/ci/azure-pipelines/try.yml
+++ b/src/ci/azure-pipelines/try.yml
@@ -28,6 +28,8 @@ jobs:
       dist-x86_64-linux: {}
       dist-x86_64-linux-alt:
         IMAGE: dist-x86_64-linux
+      dist-android: {}
+      arm-android: {}
 
 # The macOS and Windows builds here are currently disabled due to them not being
 # overly necessary on `try` builds. We also don't actually have anything that
```

**Comment by badboy** [comment]
Done!

**Comment by XAMPPRocky** [comment]
RE: Upgrading Android; I don't see any reason to not upgrade, but we should try to define a band of what API level to support. Someone who has more experience with the Android ecosystem would have to chime in here. It's my understanding that Google's supported versions doesn't necessarily line up with what being used by vendors.

**Comment by badboy** [comment]
Gnaaah, formatting error, because now that it's one item less it should be on a single line.

**Comment by ollie27** [comment]
I think it would be also be worth updating https://forge.rust-lang.org/release/platform-support.html to include the minimum supported API level.

The workarounds for missing `ftruncate64`, `pread64` and `pwrite64` can presumably be removed now as they all seem to be available in API level 16.

**Comment by badboy** [comment]
> I think it would be also be worth updating [forge.rust-lang.org/release/platform-support.html](https://forge.rust-lang.org/release/platform-support.html) to include the minimum supported API level.
> 
> The workarounds for missing `ftruncate64`, `pread64` and `pwrite64` can presumably be removed now as they all seem to be available in API level 16.

Happy to tackle that, once this actually sticks.

**Comment by nikomatsakis** [comment]
I don't see any reason not to do it, but I'm not that familiar with android. It'd be great if we had some go-to folks to ping on these matters.

**Comment by badboy** [comment]
```
couldn't find required command: "/android/ndk/arm-14/bin/arm-linux-androideabi-clang"
```

I missed an `arm-14` ...

**Comment by badboy** [comment]
@bors try

(I assume I can't actually do that)

**Comment by Mark-Simulacrum** [comment]
I think we can proceed with the android SDK api level bump here; I don't think we have any experts to consult at this point unfortunately, and if Firefox has been okay for 3 years now with this higher level of compat, we can probably be too:

> Android 4.0 and earlier versions: In September 2017, starting with Firefox version 56, only Android 4.1 and higher will be supported. If you are on Android 4.0 (Ice Cream Sandwich) and earlier, we recommend upgrading to a more recent version of Android.

from https://support.mozilla.org/en-US/kb/will-firefox-work-my-mobile-device

@badboy -- can you roll back the try builder changes and remove any of the now-unnecessary cfg's and such? Once that's done, I feel comfortable going ahead and approving, though I'll likely give it a few more days -- maybe Monday -- just to let people weigh in if they do have thoughts.

**Comment by badboy** [comment]
I'm working on on some Mozilla stuff and so far everything Android 5+ is what we support, so Android 4.1 is even more conservative.

I removed the commit for  the try builder changes.

I started looking into the other cfg'd Android stuff, but that might take me a bit more to go through fully. I'm also not sure I have everything set up to test my changes locally for that.
IMO removing these cfg'd changes could land afterwards in a separate PR, but I'm happy to adopt as you want it

**Comment by Mark-Simulacrum** [comment]
Yeah, it can definitely land later. But if there's obviously correct things, we can do that right away.

**Comment by badboy** [comment]
I added 2 commits
* one to cleanup what I think can go now
* one to revert the previously disabled tests.

I managed to compile for `i686-linux-android` locally, though I did not run tests.
I'm not 100% about the changes and I'm happy to take out either or all of them for now, and keep working on that afterwards.

**Comment by Mark-Simulacrum** [comment]
I think they all look reasonable, so I'll approve this Monday if there's no further feedback.

**Comment by goolic** [comment]
I actually would go even higher because those first API's were pretty bad. But some perspective:

We are at [2.5 billion android users](https://www.androidpolice.com/2019/05/07/there-are-now-more-than-2-5-billion-active-android-devices/), thus the 1.8% of devices using API 10, 15 and 16 stand at 45 million.

**Comment by nikomatsakis** [comment]
I approve of moving forward.

**Comment by Mark-Simulacrum** [comment]
@bors r+

Okay, let's get this kicked off. @badboy -- would you be up for filing a PR against https://github.com/rust-lang/rust-forge/ to update the compatibility table there?

**Comment by badboy** [comment]
> @bors r+
> 
> Okay, let's get this kicked off. @badboy -- would you be up for filing a PR against [rust-lang/rust-forge](https://github.com/rust-lang/rust-forge/) to update the compatibility table there?

Will do!

**Comment by badboy** [comment]
Guess I _did_ break something. Will take a look.

File: src/libstd/sys/unix/fd.rs [hunk 1]
```diff
@@ -117,10 +113,6 @@ impl FileDesc {
     }
 
     pub fn write_at(&self, buf: &[u8], offset: u64) -> io::Result<usize> {
-        #[cfg(target_os = "android")]
-        use super::android::cvt_pwrite64;
-
-        #[cfg(not(target_os = "android"))]
         unsafe fn cvt_pwrite64(
```
Comment (alexcrichton) [inline src/libstd/sys/unix/fd.rs:116]: If you'll forgive the drive-by-comment, a lot of the organization in these idioms isn't necessary any longer now that the `#[cfg]` switch is gone, so you can probably clean these all up by inlining functions instead of declaring inline functions or inline blocks. Not critical of course, but just a thought!

Comment (badboy) [reply src/libstd/sys/unix/fd.rs:116]: That absolutely makes sense, drive-by comment happily taken.
First I need to get it working, then I can make it nice.

**Comment by badboy** [comment]
Why did the tests run again at all?

**Comment by Dylan-DPC** [comment]
No idea :D could be some weird queue sync

**Comment by badboy** [comment]
Ok, I'm back on this now to fix it.

**Comment by badboy** [comment]
Ok, I got a reliable build environment working now and managed to fix above problems.

However, looks like I did not do my research properly.
`posix_memalign` is still not available in API level 16, see https://github.com/android/ndk/issues/647

Any one got issues with bumping to API level 17 instead (and also upgrading the NDK to the latest revision that supports that API level)?

**Comment by Mark-Simulacrum** [comment]
Seems fine to bump to 17 as well, though I do note that seems to go above what Firefox claims to support - maybe good to check with someone and update that page? https://support.mozilla.org/en-US/kb/will-firefox-work-my-mobile-device

I know you mentioned 5+ as the support level you're aware of, but maybe that's not on all projects? I'm willing to go with that as consensus though if you don't think it's worth trying to find someone who can update/explain the Firefox support page.

**Comment by badboy** [comment]
Good point, that's for the old Firefox for Android though, which is already based on old Firefox code, so I'm pretty sure they also pin an old rustc version. I double-check though.

**Comment by badboy** [comment]
Quick update before I end work on this for today:

I now got it to compile against API level 17, with NDK rev 19.
The resulting compiler is failing to link for the `i686-linux-android` target due to this:

```
  = note: /home/jer/code/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/i686-linux-android/lib/libbacktrace_sys-7b99c6a3f4b1e71f.rlib(mmapio.o):mmapio.c:function __rdos_backtrace_get_view: error: undefined reference to 'getpagesize'
```

That `getpagesize` is used in [libbacktrace's mmapio.c](https://github.com/rust-lang/libbacktrace/blob/1b8128cf56f005fce2dee155f7ec0c10132be8b6/mmapio.c#L63).

According to [the Android bionic status page](https://android.googlesource.com/platform/bionic/+/master/docs/status.md) `getpagesize` is properly introduced in API level 21.
There it seems to be [hardcoded to a constant](https://android.googlesource.com/platform/bionic/+/master/libc/bionic/getpagesize.cpp#31).
But both the comment on that and the [`getpagesize` man page](http://man7.org/linux/man-pages/man2/getpagesize.2.html) say `sysconf(_SC_PAGESIZE)` is a more portable way to express this.

There's multiple ways to "fix" this:

1. Overwrite the symbol somewhere (best in backtrace-sys I guess):

```rust
#[cfg(target_os = "android")]
#[no_mangle]
pub extern "C" fn getpagesize() -> libc::c_int {
    unsafe { libc::sysconf(libc::_SC_PAGESIZE) as libc::c_int }
}
```

_(Look, [I'm a fan of these shenanigans](https://fnordig.de/2020/05/02/rust-in-an-instant/), ok?)_

2. Patch `libbacktrace` to use `sysconf`

3. Compile `libbacktrace` to use [read.c](https://github.com/rust-lang/backtrace-rs/blob/688e20e019baa5aea14947b14455e65fa725fc58/crates/backtrace-sys/build.rs#L42) instead.

cc @alexcrichton as maintainer of backtrace-rs you might have an opinion on this.

**Comment by alexcrichton** [comment]
I'd prefer to avoid introducing a symbol since it'd go into libstd which is a bit too broad. I think, though, that either a patch to libbacktrace to use `sysconf` or switching to read.c is fine.

**Comment by badboy** [comment]
> I'd prefer to avoid introducing a symbol since it'd go into libstd which is a bit too broad

:D Kinda expected (and I was not 100% serious on that one)

**Comment by badboy** [comment]
I spent some more time on this as I was wondering how backtrace compiled in the past. 
I stumbled upon [unified headers](https://android.googlesource.com/platform/ndk.git/+/master/docs/UnifiedHeaders.md) and it turns out that due to this the compiler behaves differently:

```
$ ndk/android-ndk-r19c/x86-17/bin/i686-linux-android-gcc -E android-api.c --target=i686-linux-android
[...]
APIVERSION __ANDROID_API__
$  ndk/android-ndk-r19c/x86-17/bin/i686-linux-android-gcc -E android-api.c
[...]
APIVERSION 17
```

That's because the compilers are now cross-compilers as well, so specifying the target resets the defaults.
I suppose rustc itself could (somehow) at the right place define __ANDROID_API__ as the minimum version it compiles for and get that passed through to its dependencies.
However other Rust code will eventually need to find its own way if it wants a higher level.

**Comment by badboy** [comment]
Oh, and btw: if you include any header and `__ANDROID_API__` is unspecified it is set to `10000`. That way I guess all APIs are _activated_ (but still gated by ` __attribute__((annotate("introduced_in=" "21")))` and similar)

**Comment by badboy** [comment]
I now pushed the code that I managed to build for `i686-linux-android`. It compiles a simple hello world rust now.

I also managed to run `src/ci/docker/run.sh arm-android` (after some adjustements to make it work on my machine). It took a while (just over an hour), but ultimatively succeeded. [Full log in a Gist](https://gist.github.com/badboy/be2e5c9fcf126bcbebad05644eca3c9d).

**Comment by Mark-Simulacrum** [comment]
This looks good to me -- could you update the PR description as well? With that done I think we should be good to go

**Comment by badboy** [comment]
Done.
Is it ok to keep the commits as is or want me to squash them?

I'm also happy to take some followups in cleaning up the code a bit.
Depending on when you want to land this I might get to using my locally built compiler to build our Android work as another test (might get to it on the weekend).

**Comment by Mark-Simulacrum** [comment]
Commits look fine to me.

I'm going to go ahead and approve this, I think followups can land in subsequent PRs. If we come up with any problems during testing and such we can also always revert, the patch is not too large.

@bors r+ rollup=never

**Comment by badboy** [comment]
> cp: cannot stat 'C:/hostedtoolcache/windows/Python/3.7.6/x64/python.exe': No such file or directory

uhm...wat

**Comment by badboy** [comment]
Same error on other PRs (e.g. #70733)

**Comment by JohnTitor** [comment]
That failure is unrelated to this PR.
@bors retry

**Comment by kennytm** [comment]
@bors r-

[arm-android](https://github.com/rust-lang-ci/rust/runs/656206484) still failed on #55861 😕.

```
died due to signal 11
error: test failed, to rerun pass '-p alloc --test collectionstests'
```

**Comment by badboy** [comment]
Damn, guess we can leave these tests deactivated then and check them in a followup again.

**Comment by Mark-Simulacrum** [comment]
Hm, if we don't fix the tests, are there other benefits to doing this? Just inducing (admittedly theoretical) breakage seems not great to me, if we're not getting anything out of it.

**Comment by badboy** [comment]
I guess not much more for now, other than a little bit of (possible) cleanup of the code.

When I'm back at work next monday I can take another stab at these tests and why they keep failing on the upgraded SDK.

**Comment by Dylan-DPC** [comment]
@badboy any updates?

**Comment by badboy** [comment]
Unfortunately not. I have not yet had the time to re-run the tests to reproduce the issue first.

**Comment by Dylan-DPC** [comment]
I'm closing this due to inactivity.
